### ISSUE TITLE
Potential race-conditions -> Duplicated Invoice Numbers ???

### ISSUE THREAD

**Issue by jedie** [django-oscar/django-oscar-invoices#17]
Think there is a potential problem: Invoices with the same numbers could be generated.

**oscar_invoices.utils.InvoiceCreator** does this:
```
    def generate_invoice_number(self, **kwargs):
        Invoice = self.get_invoice_model()
        pk = Invoice.get_last_pk() + 1
        return '%06d' % pk
```

`last pk +1` doesn't guarantee that duplicate invoice numbers will not be created when a next invoice is generated before the previous one is saved.

Solution could be to lock the table. Or else: Create a new Invoice instance first and then use the pk from the created instance, isn't it?

**Comment by jedie** [comment]
I also see that **Invoce.number** is not marked with "unique=True" either! This would at least prevent duplicate entries from being saved.

**Comment by jedie** [comment]
No response since 14 days... Is this project alive?

**Comment by sasha0** [comment]
Yes, the project is alive.

Would you provide a PR with the suggested changes?

**Comment by sasha0** [comment]
> I also see that Invoce.number is not marked with "unique=True" either! This would at least prevent duplicate entries from being saved.

Fixed in e7c46763285df0dd1f9cdfc2c30e714e09865cdc

**Comment by sasha0** [comment]
Thanks for the report.

We addressed it in the PR #22.
